Integration test docs (#12248)
sourcegraph/sourcegraph: doc/dev/testing.md

You can find a complete list of all possible options in the [Mocha documentation](https://mochajs.org/#command-line-usage).

#### Troubleshooting failing browser-based tests

Some common failure modes:

- Timed out waiting for http://localhost:7080 to be up: the `sourcegraph/server` container failed to start, so check the container logs that appear further down in the Buildkite output.
- Timed out waiting for a selector to match because the CSS class in the web app changed: update the test code and implementation if the CSS selector is not a stable `test-*` identifier.
- Timed out waiting for a selector to match because the page was still loading: use `waitForSelector(selector, { visible: true })`.
- Page disconnected or browser session closed: another part of the test code might have called `page.close()` asynchronously, the browser crashed (check the video), or the build got canceled.
- Node was detached from the DOM: components can change the DOM asynchronously, make sure to not rely on element handles.
- Timing problems: Use `retry()` to "poll" for a condition that cannot be expressed through `waitForSelector()` (as opposed to relying on a fixed `setTimeout()`).

> Retrying the Buildkite step can help determine whether the test is flaky or broken. If it's flaky, disable it with `it.skip()` and file an issue on the author.

Reviewer: ```suggestion
Retrying the Buildkite step can help determine whether the test is flaky or broken. If it's flaky, [disable it with `it.skip()` and file an issue on the author](https://about.sourcegraph.com/handbook/engineering/testing#flaky-tests).
```

All backend interactions are stubbed or recorded and replayed.
The integration test suite for the webapp can be found in [`web/src/integration`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/web/src/integration).

Test coverage from integration tests is tracked in [Codecov](https://codecov.io/gh/sourcegraph/sourcegraph) under the flag `integration`.

#### Running client integration tests

Run `yarn test-integration` in the repository root.
A Sourcegraph instance does not need to be running, because all backend interactions are stubbed.

See the above sections for how to debug the tests, which applies to both integration and end-to-end tests.

#### Writing integration tests

Just like end-to-end tests, integration tests use the [test driver](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/shared/src/testing/driver.ts#L129:17) which is created for each test in a `before()` hook.
In opposite to end-to-end tests, integration tests do not need to set up any backend state.
Instead, integration tests create a _test context_ object before every test using `beforeEach()`, which manages the mocked responses.

##### Mocking GraphQL responses

> Calling `testContext.overrideGraphQl()` in a test or `beforeEach()` hook with an object map allows you to override GraphQL queries and mutations that are made by the application during the test by their unique query name used by the implementation.

Reviewer: ```suggestion
Calling `testContext.overrideGraphQl()` in a test or `beforeEach()` hook with an object map allows you to override GraphQL queries and mutations made by the client code. The map is indexed by the unique query name specified in the implementation, for example [`ResolveRepo`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@a3b40f3ae9376b42ce9a67b5a33f177ba98ac050/-/blob/browser/src/shared/repo/backend.tsx?subtree=true#L32-36).
```

> The types of each query are specifically generated for each query to validate the shape of the mock results and provide autocompletion.

Reviewer: ```suggestion
Types for the variables and results of each query are generated, to validate the shape of the data at compile time and provide autocompletion.
```

Author: I wanted to express here that there is one type per query (now), as opposed to our previous one-size-fits-all types. Any suggestion to better communicate that?

Author: Ah. How about replacing the first "each query" with "the overrides"?
```suggestion
The TypeScript types of the overrides are specifically generated for each query to validate the shape of the mock results and provide autocompletion.
```

Reviewer: Yup, that works.

> If a query is made during a test that is not mocked, the request will be rejected and an error will be logged with details about which query was missing.

Reviewer: ```suggestion
If, during a test, a query is made that has no corresponding mock, the request will be rejected and an error will be logged with details about the query.
```

There are default mock responses for queries made in almost every test, which you can extend with object spread syntax if needed.

##### Waiting for a mutation and checking passed variables

> To test the client made a GraphQL mutation to the backend, you can use `waitForRequest()`.

Reviewer: ```suggestion
To verify that the client sent a GraphQL mutation to the backend, you can use`waitForRequest()`.
```

Pass it a callback that triggers the request (e.g. clicking a "Save" button in a form).
The function returns the variables that were passed to the mutation, which can be asserted with `assert.deepStrictEqual()`.

Only use `waitForRequest()` for behavior you need to test, not to generally wait for parts of the application to load.
Whether a query is made for loading is an implementation detail, instead assert and wait on the DOM using `waitForSelector()` or `retry()`.

##### Mocking JSContext

The backend provides the webapp with a context object under `window.context`.
You can override this object in integration tests using `testContext.overrideJsContext()`.
There is a default mock JSContext that you can extend with object spread syntax if needed.

### End-to-end tests

End-to-end tests test the whole app: JavaScript, CSS styles, and backend.
They can be found in [`web/src/end-to-end`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/web/src/end-to-end).

> A special end-to-end test suite is the regression test suite, which was created specifically for release testing and also contains some manual verification steps. As part of moving most of our current end-to-end tests to client & backend integration tests, the regression test suite will gradually be trimmed and phased out.

Reviewer: ```suggestion
The **regression test suite** is a special end-to-end test suite, which was created specifically for release testing and also contains some manual verification steps. As part of moving most of our current end-to-end tests to client & backend integration tests, the regression test suite will gradually be trimmed and phased out.
```

Then you can select the button with `[data-test-item-name="foo"] .test-item-delete-button`.

#### Element references

It's generally unreliable to hold references to items that are acted upon later.
In other words, don't do this:

```ts
const elem = page.selector('.selector')
elem.click()
```

Do this:

```ts
page.click('.selector')
```

> For more complex interactions, you can execute them atomically _within the browser_ using `page.evaluate()`.

Reviewer: ```suggestion
You can execute more complex interactions atomically _within the browser_ using `page.evaluate()`.
```
